vetiver/meta.py

```diff
@@ -1,23 +1,27 @@
-def _model_meta(
-    user: dict = None, version: str = None, url: str = None, required_pkgs: list = None
-):
-    """Populate relevant metadata for VetiverModel
+from dataclasses import dataclass
 
-    Args
-    ----
-        user: dict
-           Extra user-defined information
-        version: str
-            Model version, generally populated from pins
-        url: str
-            Discoverable URL for API
-        required_pkgs: list
-            Packages necessary to make predictions
-    """
-    meta = {
-        "user": user,
-        "version": version,
-        "url": url,
-        "required_pkgs": required_pkgs,
-    }
-    return meta
+
+@dataclass
+class VetiverMeta:
+    """Metadata in a VetiverModel"""
+
+    user: "dict | None" = None
+    version: "str | None" = None
+    url: "str | None" = None
+    required_pkgs: "list | None" = None
+
+    @classmethod
+    def from_dict(cls, metadata, pip_name, pkg) -> "VetiverMeta":
+
+        if metadata:
+            user = metadata.get("user", metadata)
+            version = metadata.get("version", None)
+            url = metadata.get("url", None)
+            required_pkgs = metadata.get("required_pkgs", [])
+        else:
+            user, version, url, required_pkgs = None, None, None, []
+
+        if not list(filter(lambda x: pip_name in x, required_pkgs)):
+            required_pkgs = required_pkgs + [f"{pip_name}=={pkg.__version__}"]
```

Comment:
We outlined the approach we're going to try to take in our 1-on-1 doc; it might be a good idea to write this out somewhere in both repos. Like you are saying, it turns out there are different options/norms in R and Python.

Both start out with a base level of just the packages directly required to make a prediction. This is some level of likely to work and possibly be enough, especially in R where updating to latest is basically always the right move. Then both R and Python will have an option to escalate to more robust package version tracking.

- In R, we're going straight to renv, since that is the tool most people are familiar with for this type of task, a tool we have input into how it develops, etc. So there are two levels, both familiar to R users: only package names, plus opt in to full renv.
- In Python, the thing that is most equivalent to renv (`pipfile.lock`) can seem like overkill and may be less familiar to many practitioners. Instead we can use piptools to generate a `requirements.txt` that is pinned to specific versions and covers the whole dependency graph. So there are two levels here too, but they are different to be more comfortable for Python users: only package names, plus opt in to the piptools pinned requirements.

Did I summarize that correctly @isabelizimm? Anything else to add?

Reply:
Thanks -- that's helpful context! Can you say a bit more about why a version is pinned in required_pkgs for python? Is it so piptools can generate a requirements.txt? 

I wonder in that case, if [vetiver attach_pkgs](https://github.com/rstudio/vetiver-python/blob/main/vetiver/attach_pkgs.py#L19) should get the current versions of all dependencies it is using when generating requirements.txt? For example, when generating a Dockerfile, pin using the current versions of vetiver, `<required_pkgs>`, fastapi, etc..

It looks like [this is what vetiver-r does](https://github.com/rstudio/vetiver-r/blob/main/R/write-docker.R#L107), but vetiver-python only pins `required_pkgs` and `vetiver` (w/o version specified) by default.

You can get the version of a package using this code...

```python
from importlib_metadata import version

version("vetiver")
```

Reply:
Ah -- vetiver has things like fastapi and rsconnect listed as dependencies, so what is happening when you prepare_docker is...

* load_pkgs uses piptools to generate requirements.txt from `required_pkgs` and `"vetiver"`
* because vetiver has fastapi as a dependency, both these get included
* since vetiver does not have a version specified, their versions are not the ones installed, but the latest (compatible with `required_pkgs`?)

I'm not familiar enough with renv, is vetiver-r pinning the version of vetiver installed, or the latest version of vetiver?

Reply:
> I'm not familiar enough with renv, is vetiver-r pinning the version of vetiver installed, or the latest version of vetiver?

It records the versions installed when you make the snapshot, so the version installed (not necessarily latest).

Reply:
recorded this in #140